src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticAnalyzerService_IncrementalAnalyzer.cs

```diff
@@ -17,7 +20,19 @@ namespace Microsoft.CodeAnalysis.Diagnostics
     internal partial class DiagnosticAnalyzerService : IIncrementalAnalyzerProvider
     {
         public IIncrementalAnalyzer CreateIncrementalAnalyzer(Workspace workspace)
-            => _map.GetValue(workspace, _createIncrementalAnalyzer);
+        {
+            if (GlobalOptions.IsPullDiagnostics(InternalDiagnosticsOptions.NormalDiagnosticMode))
```

Comment:
`CreateIncrementalAnalyzer` API is invoked only once per each `IIncrementalAnalyzerProvider` for the entire VS session. So, this would mean that if the user changes the Pull Diagnostics option, they would be forced to restart VS before seeing the new behavior. If we are fine with restart required, then this is a reasonable change. However, if we want the new behavior to kick in the same VS session you can take the following approach:

1. Add a new property, say `IsDisabled` to `IIncrementalAnalyzer`
2. Default this to `false` in `IncrementalAnalyzerBase` and update `DiagnosticIncrementalAnalyzer` to rely on this property depending on pull diagnostics option
3. Check the `IsDisabled` property at the start of the core loop in solution crawler that calls into incremental analyzers. All calls into incremental analyzers should go through the below `RunAnalyzersAsync` method: https://github.com/dotnet/roslyn/blob/5b9a9d44eee29f433d6c9ef9740642dd4d08e963/src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.IncrementalAnalyzerProcessor.cs#L229-L242

Reply:
> So, this would mean that if the user changes the Pull Diagnostics option, they would be forced to restart VS before seeing the new behavior. If we are fine with restart required, then this is a reasonable change. 

This is already true as we can't easily change the LSP server capabilities after it has already started.  So I think a restart is OK.  But thanks for showing the other otpion!